tests/Doctrine/Tests/ORM/Functional/Ticket/GH7820Test.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\StringType;
+use Doctrine\DBAL\Types\Type;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use function array_map;
+use function iterator_to_array;
+
+/**
+ * @group GH7820
+ *
+ * When using a {@see \Doctrine\ORM\Tools\Pagination\Paginator} to iterate over a query
+ * that has entities with a custom DBAL type used in the identifier, then `$id->__toString()`
+ * is used implicitly by {@see \PDOStatement::bindValue()}, instead of being converted by the
+ * expected {@see \Doctrine\DBAL\Types\Type::convertToDatabaseValue()}.
+ *
+ * In order to reproduce this, you must have identifiers implementing
+ * `#__toString()` (to allow {@see \Doctrine\ORM\UnitOfWork} to hash them) and other accessors
+ * that are used by the custom DBAL type during DB/PHP conversions.
+ *
+ * If `#__toString()` and the DBAL type conversions are asymmetric, then the paginator will fail
+ * to find records.
+ *
+ * Tricky situation, but this very much affects `ramsey/uuid-doctrine` and anyone relying on (for
+ * example) the {@see \Ramsey\Uuid\Doctrine\UuidBinaryType} type.
+ */
+class GH7820Test extends OrmFunctionalTestCase
+{
+    private const SONG = [
+        'What is this song all about?',
+        'Can\'t figure any lyrics out',
+        'How do the words to it go?',
+        'I wish you\'d tell me, I don\'t know',
+        'Don\'t know, don\'t know, don\'t know, I don\'t know!',
+        'Don\'t know, don\'t know, don\'t know...',
+    ];
+
+    protected function setUp() : void
+    {
+        parent::setUp();
+
+        if (! Type::hasType(GH7820LineTextType::class)) {
+            Type::addType(GH7820LineTextType::class, GH7820LineTextType::class);
+        }
+
+        $this->setUpEntitySchema([GH7820Line::class]);
+
+        foreach (self::SONG as $index => $line) {
+            $this->_em->persist(new GH7820Line(GH7820LineText::fromText($line), $index));
+        }
+
+        $this->_em->flush();
+    }
+
+    public function testWillFindSongsInPaginator() : void
+    {
+        $query = $this->_em->getRepository(GH7820Line::class)
+            ->createQueryBuilder('l')
+            ->orderBy('l.lineNumber', Criteria::ASC);
+
+        self::assertSame(
+            self::SONG,
+            array_map(static function (GH7820Line $line) : string {
+                return $line->toString();
+            }, iterator_to_array(new Paginator($query)))
+        );
+    }
+}
+
+/** @Entity */
+class GH7820Line
+{
+    /**
+     * @var GH7820LineText
+     * @Id()
+     * @Column(type="Doctrine\Tests\ORM\Functional\Ticket\GH7820LineTextType")
```

Comment:
Yeap, will need to re-visit this, since it didn't like it last time I tried it.

Reply:
Not applicable: seems like I hit an annotation bug in this case, since the class is ignored (not part of `use` statements, after all.

Won't fix for this patch though.

Reply:
That's quite unexpected! Well, nothing we can sort out now :smile: